core/src/nodes/listeners.rs

```diff
@@ -83,7 +86,22 @@ where
     /// Transport used to spawn listeners.
     transport: TTrans,
     /// All the active listeners.
-    listeners: VecDeque<Listener<TTrans>>
+    listeners: VecDeque<Listener<TTrans>>,
+    /// The next listener ID to assign.
+    next_id: ListenerId
+}
+
+/// The ID of a single listener.
+///
+/// It is part of most [`ListenersEvent`]s and can be used to remove
+/// individual listeners from the [`ListenersStream`].
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct ListenerId(u64);
+
+impl fmt::Display for ListenerId {
```

Comment:
Shouldn't that just be `Debug`?

Reply:
You mean, `ListenerId` should not have a `Display` impl?

Reply:
Yes

Reply:
Why not?

Reply:
More like: why? As far as I know, `Display` is for types that you may want to show to the end-user of your application. `Debug` is for types that you intend to show to the programmer.

Reply:
When it comes to implementing `std` traits "Why not?" is actually the better question to ask I think. We also derive a couple of `std` traits like `PartialOrd` et al. which we do not use ourselves but as a public API type we should ask why we want to prevent users to use `ListenerId` in some particular context. Why should we make it more difficult than necessary to display `ListenerId` to users? `std::fmt` states in [fmt::Display vs fmt::Debug](https://doc.rust-lang.org/std/fmt/index.html#fmtdisplay-vs-fmtdebug):

> `fmt::Display` implementations assert that the type can be faithfully represented as a UTF-8 string at all times.

which is certainly true for `ListenerId` and the [Rust API Guidelines](https://rust-lang-nursery.github.io/api-guidelines/interoperability.html) recommend:

> [...] crates that define new types should eagerly implement all applicable, common traits

Therefore I think implementing `Display` is fair.

Reply:
But the internal representation of `ListenerId` is at the discretion of `listeners.rs`. It could change in the future.

In other words, a `ListenerId` is not a defined as being a number, but as being an opaque identifier. And opaque identifiers can't be faithfully represented as a UTF-8 string.

Reply:
> But the internal representation of `ListenerId` is at the discretion of `listeners.rs`. It could change in the future.

Sure it can. But we make several commitments, for instance that a `ListenerId` can be compared for equality, can be ordered, can be copied or cloned, a hash value of it can be computed and that it has a UTF-8 representation. I do not understand why you take issue with the latter but none of the former commitments.

> In other words, a `ListenerId` is not a defined as being a number, but as being an opaque identifier that is valid in a certain context. And opaque identifiers can't be faithfully represented as a UTF-8 string.

Most identifiers, opaque or not, have UTF-8 representations, e.g. UUID, IBAN, etc. In fact a Version-4 UUID is a truly opaque identifier with a UTF-8 representation whereas  we make no effort to hide metadata and can hardly say that `ListenerId` is opaque. And while it is not itself a number it is nonetheless represented as a number.

All that being said I do not want to spend more time arguing about such a minor point and will therefore remove the `Display` impl of `ListenerId`.